### PR TITLE
#3815 - Tookan Actions

### LINKED ISSUE #3815: Tookan
### App
Tookan

### Actions
* Create an Appointment Task
* Create a Delivery Task
* Create an FOS Workflow Task
* Create a Pick Up Task
* Create a Pick and Delivery Task

### PR DESCRIPTION
**Pull request by GTFalcao**
# New Actions
- **Create Appointment Task**
- **Create Delivery Task**
- **Create Field Workforce Task**
- **Create Pickup Task**
- **Create Pickup and Delivery Task**

Closes #3815 

<a href="https://gitpod.io/#https://github.com/PipedreamHQ/pipedream/pull/4129"></a>

### CODE REVIEW

**Comment by dylburger** [comment]
Thanks for submitting this PR! When we review PRs, we follow the [Pipedream component guidelines](https://pipedream.com/docs/components/guidelines/). If you're not familiar, here's a quick checklist:

- [ ] [Create components to address specific use cases whenever possible](https://pipedream.com/docs/components/guidelines/#component-scope)
- [ ] [Component `key`s should follow the format `app_name_slug-slugified-component-name`](https://pipedream.com/docs/components/guidelines/#component-key-pattern)
- [ ] Components should follow the [standard directory structure](https://pipedream.com/docs/components/guidelines/#folder-structure)
- [ ] [Prefer Node.js client libraries to REST APIs](https://pipedream.com/docs/components/guidelines/#using-apis-vs-client-libraries)
- [ ] When making API requests, [handle pagination](https://pipedream.com/docs/components/guidelines/#pagination) to ensure all data / events are processed
- [ ] [Use `secret` props](https://pipedream.com/docs/components/guidelines/#capturing-sensitive-data) to capture sensitive data
- [ ] Props and methods should be defined in [app files](https://pipedream.com/docs/components/guidelines/#app-files) whenever possible
- [ ] Document methods with [JS Docs](https://pipedream.com/docs/components/guidelines/#methods)
- [ ] [Use `optional` props](https://pipedream.com/docs/components/guidelines/#optional-vs-required-props) whenever possible, and [set a `default` value](https://pipedream.com/docs/components/guidelines/#default-values) where you can
- [ ] Use [async options](https://pipedream.com/docs/components/guidelines/#async-options) to accept user input wherever possible

File: components/tookan/actions/create-appointment-task/create-appointment-task.ts [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+        job_delivery_datetime: this.jobDeliveryDatetime,
+        job_pickup_datetime: this.jobPickupDatetime,
+        has_delivery: 0,
+        has_pickup: 0,
+        layout_type: 1,
+      },
+    };
+    const data: object = await this.tookan.createAppointmentTask(params);
+
+    $.export("$summary", "Created appointment task successfully");
```
Comment (lcaresia) [inline components/tookan/actions/create-appointment-task/create-appointment-task.ts:51]: You can update this on others exports too.

Obs: I didn't tested this

```suggestion
    $.export("$summary", `Created appointment task successfully with id ${data.id}`);
```

File: components/tookan/app/tookan.app.ts [hunk 2]
```diff
@@ -1,13 +1,109 @@
...
+    async createTask(params: { data: object; }) {
+      const response: CreateTaskResponse = await this._httpRequest({
+        endpoint: "/create_task",
+        method: "POST",
+        ...params,
+      });
+
+      if (response.status !== 200) {
+        throw new Error("Tookan response: " + response.message);
+      }
```
Comment (lcaresia) [inline components/tookan/app/tookan.app.ts:87]: Don't need this, Pipedream axios lib will catch this automatically

```suggestion
```

**Review by lcaresia — CHANGES_REQUESTED** [review]
Thanks for your contribution!

I requested just some fixes!

File: components/tookan/app/tookan.app.ts (hunk 2 above)
Comment (GTFalcao) [reply components/tookan/app/tookan.app.ts:87]: Normally yes, but Tookan's API oddly returns 201 when there is an error [(this is documented here)](https://tookanapi.docs.apiary.io/#introduction/error-codes).

This is unusual so Pipedream does not recognize it as an error - here's an example testing the action without this condition:

image

File: components/tookan/actions/create-appointment-task/create-appointment-task.ts (hunk 1 above)
Comment (GTFalcao) [reply components/tookan/actions/create-appointment-task/create-appointment-task.ts]: Thanks for the suggestion! I've included the id on all the actions' summaries (it's returned as `job_id` from the API)

**Review by lcaresia — APPROVED** [review]
LGTM!

**Comment by DilanAthukorala** [comment]
@GTFalcao When creating tasks with additional parameters, These params are not being passed / reflected in Tookan platform. 
image

**Comment by GTFalcao** [comment]
@DilanAthukorala thanks! I've fixed this and tested it, should be working as expected now!

**Pull request merged**